core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -97,11 +101,17 @@ case class ExceptionFailure(
     description: String,
     stackTrace: Array[StackTraceElement],
     fullStackTrace: String,
-    metrics: Option[TaskMetrics])
+    metrics: Option[TaskMetrics],
+    exception: Option[Throwable] = None)
```

Comment:
@pwendell You're right - it will break pattern matching on the class. My understanding is that an `unapply` method won't help since pattern matching won't use it (they are for user code).

Case classes don't play well with binary compatibility, it seems. To do this compatibly, we'd have to have another case class, called `ExceptionFailureWithCause` say, and a trait that both it and `ExceptionFailure` extend with the common fields. Then everywhere that handles `ExceptionFailure` would also have to handle `ExceptionFailureWithCause`.

Having said all that, this class is marked `@DeveloperApi` so it's within the contract to change it. The `fullStackTrace` field was added last November, for example. I can understand the general reluctance to change code even if it is marked as being for developers only, but it's not clear if the workaround here to preserve binary compatibility is worth the complexity it adds.